File: src/bsearch/resolver.py
from __future__ import annotations

import logging
from datetime import datetime
from typing import TYPE_CHECKING

from atproto import AsyncClient, models

if TYPE_CHECKING:
    from bsearch.config import Config
    from bsearch.models import Post

logger = logging.getLogger(__name__)


class ATProtoResolver:
    """AT Protocol API client for resolving posts and backfilling history."""

    def __init__(self, config: Config) -> None:
        self.config = config
        self.client = AsyncClient(base_url=config.pds_url)
        self._logged_in = False

    async def login(self) -> None:
        """Authenticate with the AT Protocol server."""
        await self.client.login(
            login=self.config.handle,
            password=self.config.app_password,
        )
        self._logged_in = True
        logger.info("Logged in as %s", self.config.handle)

    async def _ensure_logged_in(self) -> None:
        if not self._logged_in:
            await self.login()

    async def resolve_handle(self, handle: str) -> str:
        """Resolve a handle to a DID."""
        await self._ensure_logged_in()
        response = await self.client.resolve_handle(handle)
        return response.did

    async def resolve_post_uris(self, uris: list[str]) -> list[Post]:
        """Resolve a batch of post URIs (max 25 per call) to Post objects."""

        await self._ensure_logged_in()
        if not uris:
            return []

        posts = []
        # Process in chunks of 25 (API limit)
        for i in range(0, len(uris), 25):
            chunk = uris[i : i + 25]
            try:
                response = await self.client.get_posts(chunk)
                for post_view in response.posts:
                    post = _post_view_to_post(post_view, source="like")
                    if post is not None:
                        posts.append(post)
            except Exception:
                logger.exception("Failed to resolve post URIs: %s", chunk)
        return posts

    async def backfill_own_posts(self, limit: int | None = None) -> list[Post]:
        """Fetch historical posts by the authenticated user."""

        await self._ensure_logged_in()
        posts = []
        prev_cursor = None
        cursor = None
        count = 0

        while True:
            page_limit = min(100, limit - count) if limit else 100
            response = await self.client.get_author_feed(
                actor=self.config.did or self.config.handle,
                cursor=cursor,
                filter="posts_no_replies",
                limit=page_limit,
            )

            for feed_item in response.feed:
                # Skip reposts
                if feed_item.reason is not None:
                    continue
                post = _post_view_to_post(feed_item.post, source="backfill_post")
                if post is not None:
                    posts.append(post)
                    count += 1
                    if limit and count >= limit:
                        return posts

            prev_cursor = cursor
            cursor = response.cursor
            if not cursor or cursor == prev_cursor:
                break

        return posts

    async def backfill_likes(self, limit: int | None = None) -> list[Post]:
        """Fetch historically liked posts."""
        await self._ensure_logged_in()
        posts = []
        prev_cursor = None
        cursor = None
        count = 0

        while True:
            page_limit = min(100, limit - count) if limit else 100
            response = await self.client.app.bsky.feed.get_actor_likes(
                models.AppBskyFeedGetActorLikes.Params(
                    actor=self.config.did or self.config.handle,
                    cursor=cursor,
                    limit=page_limit,
                )
            )

            for feed_item in response.feed:
                post = _post_view_to_post(feed_item.post, source="backfill_like")
                if post is not None:
                    posts.append(post)
                    count += 1
                    if limit and count >= limit:
                        return posts

            prev_cursor = cursor
            cursor = response.cursor
            if not cursor or cursor == prev_cursor:
                break

        return posts


def _post_view_to_post(post_view, source: str) -> Post | None:
    """Convert an AT Protocol PostView to our Post model."""
    from bsearch.models import Post as PostModel

    record = post_view.record
    text = getattr(record, "text", None) or ""
    if not text:
        return None

    created_at_str = getattr(record, "created_at", None)
    if created_at_str:
        try:
            created_at = datetime.fromisoformat(created_at_str.replace("Z", "+00:00"))
        except (ValueError, TypeError):
            created_at = datetime.now()
    else:
        created_at = datetime.now()

    author = post_view.author
    return PostModel(
        uri=post_view.uri,
        cid=post_view.cid,
        author_did=author.did,
        author_handle=author.handle or "",
        text=text,
        created_at=created_at,
        source=source,
    )

File: crates/bsearch-serve/src/resolver.rs
use anyhow::Context;
use anyhow::Result;
use atrium_api::agent::atp_agent::store::MemorySessionStore;
use atrium_api::agent::atp_agent::AtpAgent;
use atrium_api::app::bsky::feed::defs::PostView;
use atrium_xrpc_client::reqwest::ReqwestClient;
use bsearch_core::models::parse_created_at;
use bsearch_core::models::Post;
use bsearch_core::models::Source;

use crate::config::Config;

/// The AT Protocol caps `getPosts` at 25 URIs per call.
pub const MAX_URIS_PER_CALL: usize = 25;

/// AT Protocol client for turning liked-post references into full posts.
///
/// Mirrors the subset of `ATProtoResolver` in `src/bsearch/resolver.py` that
/// the daemon needs; backfill remains in the Python CLI.
pub struct Resolver {
    agent: AtpAgent<MemorySessionStore, ReqwestClient>,
}

impl Resolver {
    pub fn new(config: &Config) -> Self {
        Self {
            agent: AtpAgent::new(
                ReqwestClient::new(config.pds_url.clone()),
                MemorySessionStore::default(),
            ),
        }
    }

    /// Authenticate, returning the DID of the logged-in account.
    ///
    /// `createSession` already reports the account's DID, so unlike the Python
    /// code there is no need for a follow-up `resolveHandle` call.
    pub async fn login(&self, config: &Config) -> Result<String> {
        let session = self
            .agent
            .login(&config.handle, &config.app_password)
            .await
            .map_err(|e| anyhow::anyhow!("{e}"))
            .with_context(|| format!("Failed to log in as {}", config.handle))?;
        tracing::info!(handle = %config.handle, "Logged in");
        Ok(session.did.as_str().to_string())
    }

    /// Resolve liked-post URIs into posts, in chunks of [`MAX_URIS_PER_CALL`].
    ///
    /// A failing chunk is logged and skipped rather than aborting the batch,
    /// matching the Python behaviour; the caller re-queues on a hard error.
    pub async fn resolve_post_uris(&self, uris: &[String]) -> Result<Vec<Post>> {
        let mut posts = Vec::new();
        for chunk in uris.chunks(MAX_URIS_PER_CALL) {
            let params = atrium_api::app::bsky::feed::get_posts::ParametersData {
                uris: chunk.to_vec(),
            };
            match self.agent.api.app.bsky.feed.get_posts(params.into()).await {
                Ok(output) => {
                    for view in &output.posts {
                        if let Some(post) = post_view_to_post(view, Source::Like) {
                            posts.push(post);
                        }
                    }
                }
                Err(e) => {
                    tracing::error!(error = %e, uris = ?chunk, "Failed to resolve post URIs");
                }
            }
        }
        Ok(posts)
    }
}

/// Convert an AT Protocol `PostView` into our `Post`, or `None` when the
/// record carries no text (the Python code skips those too).
///
/// `record` is an untyped `Unknown`, so the text and timestamp are read out of
/// its JSON representation rather than a generated struct.
fn post_view_to_post(view: &PostView, source: Source) -> Option<Post> {
    let record = serde_json::to_value(&view.record).ok()?;
    let text = record.get("text")?.as_str()?;
    if text.is_empty() {
        return None;
    }

    let created_at = parse_created_at(record.get("createdAt").and_then(|v| v.as_str()));

    Some(Post::new(
        view.uri.clone(),
        view.cid.as_ref().to_string(),
        view.author.did.as_str().to_string(),
        view.author.handle.as_str().to_string(),
        text.to_string(),
        created_at,
        source,
    ))
}
